lib/messagebird/http_client.rb

```diff
@@ -0,0 +1,57 @@
+require 'net/https'
+require 'uri'
+
+module MessageBird
+
+  class InvalidPhoneNumberException < TypeError; end
+  class MethodNotAllowedException < ArgumentError; end
+
+  class HttpClient
+
+    attr_reader :access_key
+
+    def initialize(access_key)
+      @access_key = access_key
+    end
+
+    def request(method, path, params={})
+      uri = URI.join(ENDPOINT, '/', path)
+
+      # Set up the HTTP object.
+      http = Net::HTTP.new(uri.host, uri.port)
+      http.use_ssl     = true
+
+      # Construct the HTTP GET or POST request.
+      case method
+      when :get
+        request = Net::HTTP::Get.new(uri.request_uri)
+      when :post
+        request = Net::HTTP::Post.new(uri.request_uri)
+      else
+        raise MethodNotAllowedException
+      end
+
+      request['Accept']        = 'application/json'
+      request['Authorization'] = "AccessKey #{@access_key}"
+      request['User-Agent']    = "MessageBird/ApiClient/#{CLIENT_VERSION} Ruby/#{RUBY_VERSION}"
+
+      # If present, add the HTTP POST parameters.
+      request.set_form_data(params) if method == :post && !params.empty?
+
+      # Execute the request and fetch the response.
+      response = http.request(request)
+
+      # Parse the HTTP response.
+      case response.code.to_i
+      when 200, 201, 204, 401, 404, 405, 422
+        # Ok
+      else
+        raise InvalidPhoneNumberException, 'Unknown response from server'
+      end
+
+      response.body
```

Comment:
You do not check for content type of the response here relying that remote server returns you JSON because you asked for it in request, simply returning response's body then at the next stage `JSON.parse` throws an exception if it can't parse it.

Should you check content type, content length or actual body length here?

For example if `JSON.parse` will get an empty string at it's input it will fail with something like `unexpected token: ""` which isn't very friendly, though understandable message.

Reply:
Checking the content type sounds like a good idea! I've identified an `InvalidResponseException` and throw that in such cases. What do you think?